src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Symbols/SymbolManagerBase.cs

```diff
@@ -295,6 +295,16 @@ public TypeArray AllocParams(int ctype, CType[] prgtype)
 
         public TypeArray AllocParams(int ctype, TypeArray array, int offset)
         {
+            if (ctype == 0)
+            {
+                return s_taEmpty;
+            }
+
+            if (ctype == array.Count)
+            {
+                return array;
```

Comment:
This can be a dangerous optimization if the caller might mutate the resulting array and expect that such mutations won't affect the original.  Is that not the case here?

Reply:
The source is immutable. Also, there's an intern pool that would be searched for a matching `TypeArray` that would then find `array` and return it again, so this adds no further risk in that regard.